flask/tests/test_groups.py

```diff
@@ -0,0 +1,264 @@
+
+
+def test_list_groups_admin(test_database, client, login_as):
+    login_as("admin")
+
+    response = client.get("/api/groups")
+    assert response.status_code == 200
+    assert len(response.get_json()) == 1
+
+
+def test_list_groups_user(test_database, client, login_as):
+    login_as("user")
+
+    response = client.get("/api/groups")
+    assert response.status_code == 200
+    # Check number of groups
+    assert len(response.get_json()) == 1
+
+
+def test_list_groups_user_from_admin(test_database, client, login_as):
+    # Repeat above test from admin's eyes
+    login_as("admin")
+
+    response = client.get("/api/groups?user=2")
+    assert response.status_code == 200
+    # Check number of groups
+    assert len(response.get_json()) == 1
+
+
+# GET /api/groups/:id
+
+
+def test_get_group(test_database, client, login_as):
+    # Test invalid group_code
+    login_as("admin")
+    assert client.get("/api/groups/hahah").status_code == 403
+
+    group_2 = models.Group(group_code="alas", group_name="...")
+    db.session.add(group_2)
+    db.session.commit()
+
+    # Test wrong permissions
+    login_as("user")
+    assert client.get("/api/groups/alas").status_code == 403
+
+    # Test and validate success based on user's permissions
+    login_as("user")
+    response = client.get("/api/groups/ach")
+    assert response.status_code == 200
+    assert response.get_json()["group_name"] == "Alberta"
+
+
+# DELETE /api/groups/:id
+
+
+def test_delete_group(test_database, client, login_as, minioAdmin):
+
+    # Test without permission
+    login_as("user")
+    response = client.delete("/api/groups/haha")
+    assert response.status_code == 401
+
+    # Test with wrong code
+    login_as("admin")
+    response = client.delete("/api/groups/noCodeHere")
+    assert response.status_code == 404
+
+    # Test with user in it
+    login_as("admin")
+    response = client.delete("/api/groups/ach")
+    assert response.status_code == 422
+
+    # Standard test, need to clear users first
+    group = models.Group.query.filter_by(group_code="ach").one_or_none()
+    group.users = []
+
+    db.session.commit()
+    minioAdmin.remove_user("user")
+
+    login_as("admin")
+    assert client.delete("/api/groups/ach").status_code == 204
+
+    # Make sure it's gone
+    group = models.Group.query.filter_by(group_code="ach").one_or_none()
+    assert group == None
+    assert len(minioAdmin.list_groups()) == 0
+
+
+# PATCH /api/groups/:id
+
+
+def test_update_group(test_database, client, login_as, minioAdmin):
+    login_as("admin")
+    # Test existence
+    assert (
+        client.patch("/api/groups/code", json={"group_name": "yes"}).status_code == 404
+    )
+
+    # Test success and check db and minio
+    ad = models.User.query.filter(models.User.username == "admin").one_or_none()
+    ad.minio_access_key = "admin"
+    user = models.User.query.filter(models.User.username == "user").one_or_none()
+    user.minio_access_key = "user"
```

Comment:
This might be better for the test_database fixture instead of making use of the shared SQLAlchemy session here.

Reply:
Good point, i thought it might break other tests but it should not and does not.